src/code/z_bgcheck.c

```diff
@@ -1,461 +1,4293 @@
+            }
+            spB8.z += colCtx->unitSize.z;
+            spAC.z += colCtx->unitSize.z;
+            spA4 += sp98;
         }
     }
+    return colCtx->polyNodes.count * sizeof(SSNode);
 }
 #else
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041978.s")
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_8003BB18.s")
 #endif
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_800419B0.s")
+s16 spotScenes[19] = {
+    /* 0x51 */ SCENE_SPOT00,
+    /* 0x52 */ SCENE_SPOT01,
+    /* 0x53 */ SCENE_SPOT02,
+    /* 0x54 */ SCENE_SPOT03,
+    /* 0x55 */ SCENE_SPOT04,
+    /* 0x56 */ SCENE_SPOT05,
+    /* 0x57 */ SCENE_SPOT06,
+    /* 0x58 */ SCENE_SPOT07,
+    /* 0x59 */ SCENE_SPOT08,
+    /* 0x5A */ SCENE_SPOT09,
+    /* 0x5B */ SCENE_SPOT10,
+    /* 0x5C */ SCENE_SPOT11,
+    /* 0x5D */ SCENE_SPOT12,
+    /* 0x5E */ SCENE_SPOT13,
+    /* 0x5F */ SCENE_SPOT15,
+    /* 0x60 */ SCENE_SPOT16,
+    /* 0x61 */ SCENE_SPOT17,
+    /* 0x62 */ SCENE_SPOT18,
+    /* 0x63 */ SCENE_SPOT20,
+};
+
+// Is current scene a SPOT scene
+s32 func_8003BF18(GlobalContext* globalCtx) {
+    s16* i;
+
+    for (i = spotScenes; i < spotScenes + ARRAY_COUNT(spotScenes); i++) {
+        if (globalCtx->sceneNum == *i) {
+            return true;
+        }
+    }
+    return false;
+}
+
+struct BgChkMem {
+    s16 sceneId;
+    u32 size;
+};
 
-u32 func_80041A28(CollisionContext* a0, u32 a1, u32 a2) {
-    u32 uVar1;
+struct BgChkMem D_80119E54[8] = { { SCENE_SPOT00, 0xB798 },     { SCENE_GANON_FINAL, 0x78C8 },
+                                  { SCENE_GANON_DEMO, 0x70C8 }, { SCENE_JYASINBOSS, 0xACC8 },
+                                  { SCENE_KENJYANOMA, 0x70C8 }, { SCENE_JYASINZOU, 0x16CC8 },
+                                  { SCENE_HIDAN, 0x198C8 },     { SCENE_GANON_BOSS, 0x84C8 } };
 
-    uVar1 = func_800419B0(a0, a1, a2, 0);
-    return uVar1 & 0xff;
+// Get custom scene memSize
+s32 func_8003BF5C(s32 sceneId, u32* memSize) {
+    s32 i;
+    for (i = 0; i < 8; i++) {
+        if (sceneId == D_80119E54[i].sceneId) {
+            *memSize = D_80119E54[i].size;
+            return true;
+        }
+    }
+    return false;
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041A4C.s")
+// Compute division unit sizes for scene mesh lookup, for a single dimension
+void func_8003BFF4(f32 min, s32 subdivisions, f32* max, f32* unitSize, f32* inverseUnitSize) {
+    f32 tempUnitSize;
+    f32 length;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041B24.s")
+    length = (*max - min);
+    tempUnitSize = (s32)(length / subdivisions) + 1;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041B80.s")
+    *unitSize = tempUnitSize;
+    *unitSize = (tempUnitSize < 150.0f) ? 150.0f : *unitSize;
+    *inverseUnitSize = 1.0f / *unitSize;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041C10.s")
+    *max = *unitSize * subdivisions + min;
```

Comment:
part of the purpose of the code is to resize the lookup grid if the default CollisionHeader bounds makes unitSize too small; what's more important is that each cell has a certain minimum size, so that polys are more likely to be contained in a cell, and so that collision checks don't cross a bunch of cells.